[2단계 - HTTP 서버 구현하기] 띠용(주용은) 미션 제출합니다

jye326 opened the pull request:
다음 단계가 리팩터링이기도 하고 시간도 촉박하여 기능 구현에만 집중하였습니다!
process 클래스 로직이 복잡하여 주석을 남겨두었습니다.

자유롭게 피드백 부탁드립니다!

jye326 committed:
fix: 빌드 환경 수정

jye326 committed:
feat: FileTest 작성

jye326 committed:
feat: IOStreamTest 작성

jye326 committed:
feat: IOStreamTest 작성

jye326 committed:
빌드 설정 변경

jye326 committed:
fix: 테스트 수정

- 올바른 검증을 위해 하드코딩 된 값을 수정하였습니다.

jye326 committed:
feat: step1 기능 추가

1. GET /index.html 응답하기
2. CSS 지원하기
3. Query String 파싱

jye326 committed:
refactor: process 메서드 리팩터링

- 메서드 분리
- 에러메시지 상수화
- 유저 확인 메서드 은닉

jye326 committed:
feat: process 메서드 내부 검증 로직 추가

- uri 검증 로직 추가

jye326 committed:
feat: 컨텐츠 타입 추출 메서드 안정화

jye326 committed:
refactor: 로그인 핸들 기능 메서드 분리

jye326 committed:
refactor: 테스트 리팩터링

jye326 committed:
refactor: Http11Processor 리팩터링

- 메서드 파라미터 및 리턴 타입 개선
- `Path`와 `Files` 유틸 사용으로 리소스 처리 로직 최적화
- 정적 메서드를 인스턴스 메서드로 변경 및 코드 가독성 향상
- Http 응답 생성 로직 분리 및 메서드화

jye326 committed:
refactor: Http 상태코드 분리

jye326 committed:
refactor: Http 상태코드 분리

jye326 committed:
test: 로그인 성공 테스트 작성

jye326 committed:
Merge remote-tracking branch 'origin/step2' into step2

# Conflicts:
#	tomcat/src/main/java/org/apache/coyote/http11/HttpStatusCode.java

jye326 committed:
merge

jye326 committed:
refactor: Http11Processor

- 응답 클래스 분리

jye326 committed:
feat: 기본 응답 바디 설정

jye326 committed:
test: 로그인 성공 시 리다이랙트 테스트 작성

jye326 committed:
feat: 로그인 성공 시 리다이랙트 기능 구현

jye326 committed:
temp

jye326 committed:
Merge branch 'step2' of https://github.com/jye326/java-http into step2
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

jye326 committed:
quest

jye326 committed:
test: register test

jye326 committed:
feat: 요청 클래스 분리

jye326 committed:
feat: register 기능 추가

jye326 committed:
feat: 로그인 방식 추가

jye326 committed:
feat: 세션 관리 기능 추가

- 로그인 처리에 세션 관리 로직 추가
- Cookie 클래스 및 Session 관련 클래스 구현
- Http11Processor에 세션 로직 통합
- SessionMap을 통해 세션 저장소 관리

jye326 commented:
@coderabbitai 최대한 간단히 리뷰해줘. 사족없이 최대한 간결하게 리뷰해.

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -27,7 +32,16 @@ public class Http11Processor implements Runnable, Processor {
 
+    private Request request;
+
+    private Response response;
```
final 선언 안하신 이유가 있을까요? 🧐

jye326 replied:
request는 요청이 들어오면 값을 재할당 해야 하므로 final을 붙이지 않았습니다!
response도 요청이 들어올 때마다 새로운 객체를 생성하도록 변경하였습니다.

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -27,7 +32,16 @@ public class Http11Processor implements Runnable, Processor {
+
+    private Response response;
+
+    private SessionMap sessionMap;
```
final 선언 안하신 이유가 있을까요? 🧐

jye326 committed:
refactor: Cookie 및 Session 관련 로직 개선

- Cookie 클래스를 CookieManager로 리네이밍
- SessionMap을 SessionManager로 리네이밍 및 기능 확장
- Http11Processor에 SessionManager 통합
- Response에서 Cookie 관리를 위한 로직 분리 및 개선
- 코드 가독성 및 유지보수성을 위한 전반적인 구조 리팩터링

jye326 committed:
fix: CookieManager 로직 수정

- 빈 값 쿠키 처리 중 불필요한 로직 실행 방지 위해 return 추가

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -43,89 +57,176 @@ public void process(final Socket connection) {
-        outputStream.write(response.getBytes());
+    private boolean login(Map<String, String> params) {
+        String account = params.get("account");
+        String password = params.get("password");
```
`account` 와 `password` 는 반드시 포함하고 있을까요?

jye326 replied:
관련 예외 처리 추가했습니다!👍

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -43,89 +57,176 @@ public void process(final Socket connection) {
+        String email = params.get("email");
+        User user = new User(account, password, email);
+        InMemoryUserRepository.save(user);
+        return InMemoryUserRepository.findByAccount(account).isPresent();
```
좋은 코드네요 👍

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -43,89 +57,176 @@ public void process(final Socket connection) {
 
-    private void parseQueryString(String queryString, Map<String, String> params) {
+    private Map<String, String> parseQueryParameter(String uri) {
+        if (!uri.contains("?")) {
```
`uri.conatins("?")` 분기문 타고 내려온 것 같은데 추가 방어하는 이유가 있나요?

jye326 replied:
다른 장소에서도 쓰일 수도 있을 것 같아 예외를 던지기 위해 방어했습니다!

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -43,89 +57,176 @@ public void process(final Socket connection) {
             contentType = "text/html";
         }
-        return contentType;
+        return contentType + ";charset=utf-8";
```
null 인 경우를 제외하고 모두 붙여주는 이유가 있을까요? 🧐

jye326 replied:
아직 인코딩 정보를 따로 관리하고 있지 않고 일단 UTF-8 인코딩 응답을 하드코딩 하였습니다. 시간이 나면 리팩터링 해야 할 것 같아요!

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/SessionMap.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public Session findSession(String sessionId){
+        return sessionMap.get(sessionId);
```
없을 경우도 있을 것 같아요

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -43,89 +57,176 @@ public void process(final Socket connection) {
+            String uri = request.getUrl();
+            Path path = parsePath(uri);
+            // 로그인 처리
             if (uri.startsWith("/login")) {
```
사소한데, 가장 큰 분기를 login 으로 잡으신 이유가 궁금하네요.
저는 GET / POST 로 분기를 처리하고 있어서요.

jye326 replied:
큰 이유는 없었습니다. 미션 1-3 `http://localhost:8080/login?account=gugu&password=password으로 접속하면 로그인 페이지(login.html)를 보여주도록 만들자.` 부터 순서대로 구현하다 보니 자연스럽게 이렇게 된 것 같아요.

minjae8563 reviewed (approved):
코드 잘 짜주셔서 재밌게 읽었습니다~
반영할만한 리뷰 있다면 반영해주시면 될 것 같아요 👍

minjae8563 commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -43,89 +58,187 @@ public void process(final Socket connection) {
+                session.setAttribute("user", user); // 세션에 user 정보에 user 저장
+                sessionManager.addSession(sessionId, session);  // 세션 맵에 새 세션 등록
+                CookieManager cookieManager = new CookieManager();
+                cookieManager.addCookie("JSESSIONID", sessionId);  // 쿠키에 세션 아이디 매핑
```
위에 세션에 대한 정보를 저장한 것 같은데, cookieManager 에서 저장하는 이유가 있을까요? 🧐

jye326 replied:
응답 쿠키에 세션 id 정보를 담아서 한번에 여러 쿠키 정보를 응답할 수 있도록 하기 위함입니다.

jye326 reviewed (commented):
좋은 리뷰 감사합니다 말론~!

jye326 committed:
feat: 로그인 요청 검증 로직 추가

1. 로그인 요청 시 계정 및 비밀번호 누락 검증
2. 검증 실패 시 예외 처리 및 에러 메시지 추가
3. 코드 정리 및 가독성 개선
